File: MergeSort/merge_sort.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <list>
#include <chrono>

using namespace std;

// Function declarations
void mergeSort(vector<int>& S, vector<string>& id, int left, int right);
void merge(vector<int>& S, vector<string>& id, int left, int mid, int right);
void writeCsv(const vector<int>& S, const vector<string>& id);

int main() {
    vector<int> S;
    vector<string> idArray;
    string filePath = "dataset_20000000.csv"; // Adjust path if needed
    //string filePath = "Assingment/dataset_sample_1000.csv";
    ifstream file(filePath);
    string line;

    if (!file.is_open()) {
        cerr << "Could not open file: " << filePath << endl;
        return 1;
    }

    while (getline(file, line)) {
        stringstream ss(line);
        string numStr, idStr;

        if (getline(ss, numStr, ',') && getline(ss, idStr)) {
            S.push_back(stoi(numStr));
            idArray.push_back(idStr);
        }
    }

    file.close();

    auto start = chrono::high_resolution_clock::now();
    mergeSort(S, idArray, 0, S.size() - 1);
    auto end = chrono::high_resolution_clock::now();
    auto runtime = chrono::duration_cast<chrono::milliseconds>(end - start).count();

    cout << "Runtime: " << runtime << " ms" << endl;

    writeCsv(S, idArray);


    return 0;
}

void mergeSort(vector<int>& S, vector<string>& id, int left, int right) {
    if (left >= right) return;

    int mid = (left + right) / 2;
    mergeSort(S, id, left, mid);
    mergeSort(S, id, mid + 1, right);
    merge(S, id, left, mid, right);
}

void merge(vector<int>& S, vector<string>& id, int left, int mid, int right) {
    vector<int> L, R;
    vector<string> Lid, Rid;

    for (int i = left; i <= mid; ++i) {
        L.push_back(S[i]);
        Lid.push_back(id[i]);
    }
    for (int j = mid + 1; j <= right; ++j) {
        R.push_back(S[j]);
        Rid.push_back(id[j]);
    }

    int n1 = mid - left + 1;
    int n2 = right - mid;
    int i = 0, j = 0, k = left;

    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            S[k] = L[i];
            id[k] = Lid[i];
            i++;
        } else {
            S[k] = R[j];
            id[k] = Rid[j];
            j++;
        }
        k++;
    }

    while (i < n1) {
        S[k] = L[i];
        id[k] = Lid[i];
        i++; k++;
    }

    while (j < n2) {
        S[k] = R[j];
        id[k] = Rid[j];
        j++; k++;
    }
}

void writeCsv(const vector<int>& S, const vector<string>& id) {
    // Clear the file
    ofstream clearFile("merge_sort_" + to_string(S.size()) + ".csv", ios::out);
    if (!clearFile) {
        cerr << "Error clearing file." << endl;
        return;
    }
    clearFile.close();

    // Append data to the file
    ofstream outFile("merge_sort_" + to_string(S.size()) + ".csv", ios::app);
    for (size_t i = 0; i < S.size(); ++i) {
        if (!outFile) {
            cerr << "Error writing to file." << endl;
            return;
        }
        outFile << S[i] << "," << id[i] << "\n";
    }
    outFile.close();
}

File: MergeSort/merge_sort.java
package MergeSort;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedList;

public class merge_sort {
    
  public static void main (String args[]) { 
    ArrayList<Integer> S = new ArrayList<>();
    ArrayList<String> idArray = new ArrayList<>();
    String line;
    String filePath = "dataset_20000000.csv";
    //String filePath = "Assingment/dataset_sample_1000.csv";
    try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
        while ((line = br.readLine()) != null) {
            // Split the line by comma
            String[] values = line.split(",");
            S.add(Integer.parseInt(values[0]));
            idArray.add(values[1]);
            
        }
    } catch (IOException e) {
        e.printStackTrace();
    }


    long startTime = System.nanoTime();
    mergeSort (S, idArray); 
    long endTime = System.nanoTime();
    long runtime = endTime - startTime;
    System.out.println ("Runtime: " + (runtime / 1_000_000) + " ms");
    writeCsv(S,idArray);

  } 

  static void mergeSort (ArrayList<Integer> S, ArrayList<String> id) { 

    mergeSort (S, id, 0, S.size()-1); 

  } 


  static void mergeSort (ArrayList<Integer> S, ArrayList<String> id, int left, int right) { 

    // WRITE YOUR CODE 
    int mid = (left+right)/2;
    if (left != right){
        mergeSort (S, id, left, mid);
        mergeSort (S, id, mid + 1, right);
        merge (S, id, left, mid, right);
    }
    
  } 

  static void merge (ArrayList<Integer> S, ArrayList<String> id, int left, int mid, int right) { 

    // Create LinkedList L from S[left]...S[mid] 

    LinkedList<Integer> L = new LinkedList<>(); 
    LinkedList<String> Lid = new LinkedList<>(); 

    for (int i = 0; i < mid - left + 1; i++) 
    {
        L.add (S.get(left + i)); 
        Lid.add(id.get(left + i)); 
    }
 

    // Create LinkedList R from S[mid+1]...S[right] 

    LinkedList<Integer> R = new LinkedList<>(); 
    LinkedList<String> Rid = new LinkedList<>();

    for (int j = 0; j < right - mid; j++) 
    {
        R.add (S.get(mid + 1 + j)); 
        Rid.add(id.get(mid + 1 + j)); 
    }

    int k = left; 

    while (!L.isEmpty() &&  !R.isEmpty()){
        if (L.getFirst() < R.getFirst()){
            S.set(k, L.removeFirst());
            id.set(k, Lid.removeFirst());
            k++;
        }else{
            S.set(k, R.removeFirst());
            id.set(k, Rid.removeFirst());
            k++;
        }

    }
  
    while (!L.isEmpty()){
        S.set(k, L.removeFirst());
        id.set(k, Lid.removeFirst());
        k++;
    } 
  
    while (!R.isEmpty()){
        S.set(k, R.removeFirst());
        id.set(k, Rid.removeFirst());
        k++;
    } 
  } 

  static void writeCsv(ArrayList<Integer> S, ArrayList<String> id){
    // Write the values into the csv file
    String filename = "merge_sort_" + S.size() + ".csv";
    try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
        for (int i = 0; i < S.size(); i++) {
            writer.write(S.get(i) + "," + id.get(i));
            writer.newLine();
        }
    } catch (IOException e) {
        e.printStackTrace();
    }
  }
}
